protoc-gen-twirp/generator.go

```diff
@@ -1146,6 +1146,20 @@ func (t *twirp) generateServer(file *descriptor.FileDescriptorProto, service *de
 	t.P(`}`)
 	t.P()
 
+	// Write request body read issue
+	t.P(`// handleRequestBodyError is used to handle error when the twirp server cannot read request`)
+	t.P(`func (s *`, servStruct, `) handleRequestBodyError(ctx `, t.pkgs["context"], `.Context, resp `, t.pkgs["http"], `.ResponseWriter, msg string) {`)
+	t.P(`  if ctxErr := context.Canceled; ctxErr == ctx.Err() {`)
+	t.P(`    s.writeError(ctx, resp, twirp.NewError(twirp.Canceled, ctxErr.Error()))`)
+	t.P(`    return`)
+	t.P(`  } else if ctxErr := context.DeadlineExceeded; ctxErr == ctx.Err() {`)
+	t.P(`    s.writeError(ctx, resp, twirp.NewError(twirp.DeadlineExceeded, ctxErr.Error()))`)
+	t.P(`    return`)
```

Comment:
Done

Reply:
To clarify, I also find the `ctxErr` assignment inside the conditional un-necessarily confusing to read, and it obfuscates the final error message. A developer reading this code may be trying to figure out why they see "context canceled" in the logs, it would be better IMO if we would make that as explicit as possible. `context.Canceled.Error()` is more explicit than `ctxErr.Error()`, specially if that same line means something different on each conditional.

I would prefer if we write this in a more explicit form like this:

```go
func (s *svc2Server) handleRequestBodyError(ctx context.Context, resp http.ResponseWriter, err error, msg string) {
    if ctx.Err() == context.Canceled {
        s.writeError(ctx, resp, twirp.NewError(twirp.Canceled, "failed to read request: context canceled"))
        return
    }
    if ctx.Err() == context.DeadlineExceeded {
        s.writeError(ctx, resp, twirp.NewError(twirp.DeadlineExceeded, "failed to read request: deadline exceeded"))
        return
    }
    s.writeError(ctx, resp, twirp.WrapError(malformedRequestError(msg), err)) // original error can be unwrapped for inspection
}
```
The reason is that this is both easier to read and easier to find when searching for the error message.

I know @chetanvalagit has been going back and forth with this and may not want to keep running in circles, this is more of a question for @rhysh, and while I have a preference here, I don't think this should block the PR any longer, so I'm fine with whatever you both decide now.

Reply:
Either is fine, provided the spelling is right. Pulling this out into a specialized function helps with safety, and the tests help with safety. Those two alone are probably sufficient protection against 1/ "cancelled" vs "canceled" or 2/ copy–paste errors that turn "context canceled" into "deadline exceeded".